app/jobs/export_pool_xp_to_traction_job.rb

```diff
@@ -0,0 +1,125 @@
+# frozen_string_literal: true
+
+# Combine the samples from a Pool XP tube into a compound sample, generate a bioscan-pool-xp-tube-to-traction message
+# and submit it to RabbitMQ so that it can be forwarded to Traction by the message processor.
+ExportPoolXpToTractionJob =
+  Struct.new(:barcode) do
+    include CompoundSampleHelper
+
+    def perform
+      message_data = get_message_data(barcode)
+
+      subject_obj = configatron.amqp.schemas.subjects[:export_pool_xp_to_traction]
+      subject = subject_obj[:subject]
+      version = subject_obj[:version]
+
+      schema = get_message_schema(subject, version)
+      encoded_message = avro_encode_message(message_data, schema)
+      send_message(encoded_message, subject, version)
+    rescue StandardError => e
+      Rails.logger.error("Error exporting Pool XP tube to Traction: <#{e.message}>")
+      raise
+    end
+
+    def fetch_response(uri_str, limit = 10)
+      raise IOError, 'Too many HTTP redirects' if limit.zero?
+
+      response = Net::HTTP.get_response(URI.parse(uri_str))
+
+      case response
+        when Net::HTTPSuccess     then response
+        when Net::HTTPRedirection then fetch_response(response['location'], limit - 1)
+      else
+        response.error!
```

Comment:
Where does this `error!` method come from, out of interest? Can't see it on https://ruby-doc.org/stdlib-2.7.0/libdoc/net/http/rdoc/Net/HTTPResponse.html - must be a more generic ruby thing?

Reply:
I took this and tweaked it from a highly uprated Stackoverflow answer.  I agree `error!` doesn't appear in that documentation, but it's in the source code for the `HTTPResponse` class: https://github.com/ruby/ruby/blob/master/lib/net/http/response.rb#L274-L278  It's been marked with `#:nodoc:` which is a bit concerning, but it is a public method.  Perhaps this is a pattern throughout the HTTP code and they didn't want to repeat the documentation everywhere.